### DISCUSSION TITLE
Should useRevalidator be stable?

### DISCUSSION THREAD

**Discussion by charliematters** [remix-run/react-router#10702] [Q&A] [answered]
I've got a hook which calls `useRevalidator` and ReactQuery's `invalidateQueries` to allow any component to refresh data.

I'm currently chasing down some unnecessary re-renders, and my `useRevalidator` hook is returning a new value each time. 

The docs imply that it's a singleton under the hood, so I'd assumed that we could rely on `useRevalidator` not changing.

Have I misunderstood how this works?

**Comment by brophdawg11** [comment, marked as answer]
It will change on revalidations because it also returns `useRevalidator().state` which can change.  But we should probably memoize it internally based on `state` so the object is stable on re-renders when a revalidation is not in progress.

**Reply by charliematters** [reply]
That sounds ideal. I'm not near a computer, but does that mean I can rely on useRevalidator().revalidate being (mostly) stable currently?

**Reply by brophdawg11** [reply]
Yeah that should be stable currently

**Reply by brophdawg11** [reply]
This was stabilized in https://github.com/remix-run/react-router/issues/10706 in `6.15.0`
